linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1224,23 +1233,79 @@ private[optimizer] abstract class OptimizerCore(
-            val recordField = recordType.findField(field.name)
-            val sel = RecordSelect(newQual, field)(recordField.tpe)
-            sel.tpe match {
-              case _: RecordType =>
-                cont(PreTransRecordTree(sel, RefinedType(expectedType), cancelFun))
-              case _ =>
-                cont(PreTransTree(sel, RefinedType(sel.tpe)))
+        def default: TailRec[Tree] = {
+          resolveLocalDef(preTransQual) match {
+            case PreTransRecordTree(newQual, origType, cancelFun) =>
+              val recordType = newQual.tpe.asInstanceOf[RecordType]
+              val recordField = recordType.findField(field.name)
+              val sel = RecordSelect(newQual, field)(recordField.tpe)
+              sel.tpe match {
+                case _: RecordType =>
+                  cont(PreTransRecordTree(sel, RefinedType(expectedType), cancelFun))
+                case _ =>
+                  cont(PreTransTree(sel, RefinedType(sel.tpe)))
+              }
+
+            case PreTransTree(newQual, newQualType) =>
+              val newQual1 = maybeAssumeNotNull(newQual, newQualType)
+              cont(PreTransTree(Select(newQual1, className, field)(expectedType),
+                  RefinedType(expectedType)))
+          }
+        }
+
+        preTransQual.tpe match {
+          // Try to inline an inlineable field body
+          case RefinedType(ClassType(qualClassName), _, _) if !isLhsOfAssign =>
+            if (myself.exists(m => m.enclosingClassName == qualClassName && m.methodName.isConstructor)) {
+              /* Within the constructor of a class, we cannot trust the
+               * inlineable field bodies of that class, since they only reflect
+               * the values of fields when the instance is fully initialized.
+               */
+              default
+            } else {
+              inlineableFieldBodies(qualClassName).fieldBodies.get((className, field.name)) match {
+                case None =>
+                  default
+                case Some(fieldBody) =>
+                  val qualSideEffects = checkNotNullStatement(preTransQual)
+                  val fieldBodyTree = fieldBodyToTree(fieldBody)
+                  pretransformExpr(fieldBodyTree) { preTransFieldBody =>
+                    cont(PreTransBlock(qualSideEffects, preTransFieldBody))
+                  }
+              }
             }
+          case _ =>
+            default
+        }
+    }
+  }
+
+  private def fieldBodyToTree(fieldBody: InlineableFieldBodies.FieldBody)(implicit pos: Position): Tree = {
+    import InlineableFieldBodies.FieldBody
 
-          case PreTransTree(newQual, newQualType) =>
-            val newQual1 = maybeAssumeNotNull(newQual, newQualType)
-            cont(PreTransTree(Select(newQual1, className, field)(expectedType),
-                RefinedType(expectedType)))
+    fieldBody match {
+      case FieldBody.Literal(literal) =>
+        // explicit copy for the position
```

Comment:
Is this really what we want? For other inlining, IIUC we do not do this.

Reply:
If we don't do it, we must take the `pos` of the trees into account when checking for equality of the `FieldBody`es. Otherwise an incremental run might not produce the same source map as a clean run.

Reply:
OK, makes sense. I added positions in `FieldBody` to be able to use them.